clippy_lints/src/outer_doc_comment_bang.rs

```diff
@@ -0,0 +1,99 @@
+use clippy_utils::{
+    diagnostics::span_lint_and_sugg,
+    source::{indent_of, reindent_multiline},
+};
+use if_chain::if_chain;
+use rustc_ast::{ast::*, token::CommentKind};
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use std::fmt::Write;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Detects the use of outer doc comments (`///`, `/**`) followed by a bang (`!`).
+    ///
+    /// ### Why is this bad?
+    /// Triple-slash comments (known as "outer doc comments") apply to items that follow it.
+    /// An outer doc comment followed by a bang (i.e. `///!`) has no specific meaning.
+    ///
+    /// The user most likely meant to use an inner doc comment (`//!`, `/*!`), which
+    /// applies to the parent item (i.e. the item that this comment is contained in,
+    /// usually a module or crate).
+    ///
+    /// ### Example
+    /// In this example, the doc comment is attached to the *function*, rather than the *module*.
+    /// ```rust
+    /// pub mod util {
+    ///     ///! This module contains utility functions.
+    ///
+    ///     pub fn dummy() {}
+    /// }
+    /// ```
+    ///
+    /// Use instead:
+    /// ```rust
+    /// pub mod util {
+    ///     //! This module contains utility functions.
+    ///
+    ///     pub fn dummy() {}
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub OUTER_DOC_COMMENT_BANG,
+    suspicious,
+    "outer doc comment followed by a bang symbol"
+}
+declare_lint_pass!(OuterDocCommentBang => [OUTER_DOC_COMMENT_BANG]);
+
+const WARNING: &str = "this is an outer doc comment and does not apply to the parent module or crate";
+const HELP: &str = "use an inner doc comment to document the parent module or crate";
+
+impl EarlyLintPass for OuterDocCommentBang {
+    fn check_item(&mut self, cx: &EarlyContext<'_>, item: &Item) {
+        let mut lo_span = None;
+        let mut hi_span = None;
+        let mut sugg = String::new();
+
+        // multiline doc comments span across multiple attributes
+        // so keep pushing into a string
+        for attr in &item.attrs {
+            if_chain! {
+                if let AttrKind::DocComment(com_kind, sym) = attr.kind;
+                if let AttrStyle::Outer = attr.style;
+                if let Some(com) = sym.as_str().strip_prefix('!');
```

Comment:
maybe move it to a separate function, and then we can use `map(...).collect(...)` to built `Vec<String>`